incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -783,41 +1135,1198 @@ private void onKafkaAbort(
+
+
+                        final MutableDirectBuffer dataBuffer = bufferPool.buffer(dataSlot);
+                        dataBuffer.putBytes(0, willMessage.buffer(), willMessage.offset(), willMessage.sizeof());
+
+                        messageSlotReserved = willMessage.sizeof();
+
+                        willProducer =
+                            new KafkaProduceWillProxy(originId, routedId, this, delegate.messagesTopic, deliverAt);
+                        willProducer.doKafkaBegin(traceId, authorization, 0);
+                        if ((willMessage.flags() & 1 << MqttPublishFlags.RETAIN.value()) != 0)
+                        {
+                            willRetainProducer =
+                                new KafkaProduceWillProxy(originId, routedId, this, delegate.retainedTopic, deliverAt);
+                            willRetainProducer.doKafkaBegin(traceId, authorization, 0);
+                        }
+                    }
+                    else
+                    {
+                        doKafkaEnd(traceId, authorization);
+                    }
+                }
+            }
+        }
+
+        private void onKafkaFlush(
+            FlushFW flush)
+        {
+            final long sequence = flush.sequence();
+            final long acknowledge = flush.acknowledge();
+            final long traceId = flush.traceId();
+            final long authorization = flush.authorization();
+
+            assert acknowledge <= sequence;
+            assert sequence >= replySeq;
+
+            replySeq = sequence;
+
+            assert replyAck <= replySeq;
+
+            doKafkaEnd(traceId, authorization);
+        }
+
+        private void onKafkaWindow(
+            WindowFW window)
+        {
+            final long sequence = window.sequence();
+            final long acknowledge = window.acknowledge();
+            final int maximum = window.maximum();
+
+            assert acknowledge <= sequence;
+
+            initialAck = acknowledge;
+            initialMax = maximum;
+            state = MqttKafkaState.openInitial(state);
+
+            assert initialAck <= initialSeq;
+        }
+
+        private void doKafkaReset(
+            long traceId)
+        {
+            if (MqttKafkaState.initialOpened(state) && !MqttKafkaState.replyClosed(state))
+            {
+                state = MqttKafkaState.closeReply(state);
+
+                doReset(kafka, originId, routedId, replyId, replySeq, replyAck, replyMax, traceId);
+            }
+        }
+
+        private void doKafkaWindow(
+            long traceId,
+            long authorization,
+            long budgetId,
+            int padding,
+            int capabilities)
+        {
+            replyMax = bufferPool.slotCapacity();
+
+            doWindow(kafka, originId, routedId, replyId, replySeq, replyAck, replyMax,
+                traceId, authorization, budgetId, padding, replyPad, capabilities);
+        }
+
+        public void releaseBufferSlot(
```

Comment:
`private`?

Reply:
Feel like this should happen naturally when trying to flush the buffered data when send is complete, or else stream aborts / resets and we need to cleanup the buffer.

Reply:
As currently implemented, it appears as though we are counting `WINDOW` frames to determine when to release the buffer, `wasOpen` vs `!wasOpen`. But `WINDOW` frames can be logically split or combined into any number of frames, as long as the aggregate `seq`, `ack` are consistent, for example.

Therefore, I think we would need to treat it differently. Reacting to receiving `ack` with high enough value to release the buffer.

However, I think we can instead safely release the buffer a bit sooner, when the will message is fully delivered, perhaps on both `mqtt_messages` and `mqtt_retained` in the most complex case.

This would most easily be coded by detecting progress through the buffer independently on each stream, then notifying the delegate when complete, and having an `acquire` count on the delegate that is initialized once for each `mqtt_messages` and (optional) `mqtt_retained` stream, then decrements and does the actual `release` when the count reaches `zero`, so it doesn't matter which in order `mqtt_messages` vs `mqtt_retained` complete, or even if `mqtt_retained` is needed.

If an error occurs while delivering the will message to `mqtt_messages` and/or `mqtt_retained`, I think we would want to simply cascade that error to the `will` stream, `mqtt_messages` and `mqtt_retained`, triggering a reconnect for the `will` stream which will replay the `will` message.

Note that the `ack` to the `will` stream should be deferred until both `mqtt_messages` and optionally `mqtt_retained` (if will retained) have acknowledged.